File: python/encoding/Encoders.py
import numpy as np
import time
import Predictors
import logging
from Frames import *
from golomb import Golomb
from BitStream import BitStream

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger('root')
logger.setLevel(logging.INFO)

c_handler = logging.StreamHandler()
c_format = logging.Formatter('%(name)s - %(levelname)s - %(message)s')
c_handler.setFormatter(logging.Formatter('%(name)s | %(levelname)s ->: %(message)s'))

logger.addHandler(c_handler)
logger.propagate = False  # https://stackoverflow.com/a/19561320


class IntraFrameEncoder():
    """
    Lossless intra-frame encoder
    """

    def __init__(self, matrix, predictor):
        self.original_matrix = matrix
        self.predictor = predictor
        self.encoded_matrix = np.empty(self.original_matrix.shape)  # sighly faster

        # Golomb encoder
        self.golomb = Golomb(4)
        self.bitstream = BitStream("./encoded_park_joy_444_720p50.bin", "wb")

    def write_code(self, code):
        for bit in code:
            self.bitstream.writeBit(bit, 1)

    def setMatrix(self, new_matrix):
        self.original_matrix = new_matrix

    def encode(self):

        # TODO: ver o que é aquele K do stor
        # write header with bitstream
        self.bitstream.writeString("{}\t{}".format(self.original_matrix.shape[0], self.original_matrix.shape[1]))

        self.encoded_matrix[0, 0] = int(self.original_matrix[0, 0] - self.predictor.predict(0, 0, 0))

        for col in range(1, self.original_matrix.shape[1]):
            self.encoded_matrix[0, col] = int(self.original_matrix[0, col]) - self.predictor.predict(
                self.original_matrix[0, col - 1], 0, 0)

        for line in range(1, self.original_matrix.shape[0]):
            self.encoded_matrix[line, 0] = int(self.original_matrix[line, 0]) - self.predictor.predict(0,
                                                                                                       self.original_matrix[
                                                                                                           line - 1, 0],
                                                                                                       0)

        for line in range(1, self.original_matrix.shape[0]):
            for col in range(1, self.original_matrix.shape[1]):
                # print("line: ", line, ", col: ", col, "; original: ", self.original_matrix[line, col], ", predict: ", self.original_matrix[line, col - 1], self.original_matrix[line - 1, col], self.original_matrix[col - 1, line -1])
                self.encoded_matrix[line, col] = int(self.original_matrix[line, col]) - self.predictor.predict(
                    self.original_matrix[line, col - 1], self.original_matrix[line - 1, col],
                    self.original_matrix[line - 1, col - 1])

        for line in range(self.encoded_matrix.shape[0]):
            for col in range(self.encoded_matrix.shape[1]):
                self.write_code(self.golomb.encoded_values[self.encoded_matrix[line, col]])


class IntraFrameDecoder():
    """
    Lossless intra-frame decoder, complementing the one analogous encoder
    """

    def __init__(self, matrix, predictor):
        self.original_matrix = matrix
        self.predictor = predictor
        self.decoded_matrix = np.empty(self.original_matrix.shape)  # sighly faster
        # self.encoded = False
    
    def setMatrix(self, new_matrix):
        self.original_matrix = new_matrix
    
    def decode(self):

        # TODO: ver o que é aquele K do stor
        self.decoded_matrix[0, 0] = self.original_matrix[0, 0] + self.predictor.predict(0, 0, 0)

        for col in range(1, self.original_matrix.shape[1]):
            self.decoded_matrix[0, col] = int(self.original_matrix[0, col]) + self.predictor.predict(
                self.decoded_matrix[0, col - 1], 0, 0)

        for line in range(1, self.original_matrix.shape[0]):
            self.decoded_matrix[line, 0] = int(self.original_matrix[line, 0]) + self.predictor.predict(0,
                                                                                                       self.decoded_matrix[
                                                                                                           line - 1, 0],
                                                                                                       0)

        print("1 ", self.original_matrix.shape[1])
        for line in range(1, self.original_matrix.shape[0]):
            for col in range(1, self.original_matrix.shape[1]):
                # print("line: ", line, ", col: ", col, "; original: ", self.original_matrix[line, col], ", predict: ", self.original_matrix[line, col - 1], self.original_matrix[line - 1, col], self.original_matrix[col - 1, line -1])
                self.decoded_matrix[line, col] = int(self.original_matrix[line, col]) + self.predictor.predict(
                    self.decoded_matrix[line, col - 1], self.decoded_matrix[line - 1, col],
                    self.decoded_matrix[line - 1, col - 1])


if __name__ == "__main__":
    frame = Frame420(720, 1280, "../media/park_joy_420_720p50.y4m")

    frame.advance()
    matrix = frame.getU()
    ife = IntraFrameEncoder(matrix, Predictors.JPEG1)
    ife.encode()

    print(ife.encoded_matrix)
    print(ife.encoded_matrix.shape)

    """
    import datetime
    start = datetime.datetime.now()
    #while True:
    playing = frame.advance()
    if not playing:
        pass # break
    matrix = frame.getY()
    ife = IntraFrameEncoder(matrix, "Y", [4,4,4], Predictors.JPEG1)
    ife.encode()
    matrix = frame.getU()
    ife = IntraFrameEncoder(matrix, "U", [4,4,4], Predictors.JPEG2)
    ife.encode()
    matrix = frame.getV()
    ife = IntraFrameEncoder(matrix, "V", [4,4,4], Predictors.JPEG3)
    ife.encode()
    end = datetime.datetime.now() - start
    print("Compressed in {} s".format(end.seconds))
    """

File: cpp/src/test/Encoders.cpp
#include <stdio.h>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
#include <chrono>
#include <opencv2/opencv.hpp>
#include "Frame.cpp"
#include "BitStream.cpp"
#include "Golomb.cpp"
#include "Predictor.cpp"


using namespace cv;

class IntraFrameEncoder
{
    private:
        std::vector<int> original_matrix;
        std::vector<int> encoded_matrix;

    public:
        IntraFrameEncoder() {

        }

        void writeCode(std::vector<int> code){};

        void setMatrix(std::vector<int> matrix)
        {   
            for(int i = 0; i < matrix.size(); i++) {
                this->original_matrix.push_back(matrix.at(i));
                
            }
        }

        void encode(Frame* frame, Predictor* predictor, BitStream* bitstream, Golomb* golomb) {
            
            int height = frame->getHeight(), width = frame->getWidth();

            int firstValue = this->original_matrix.at(0) - predictor->predict(0,0,0);
            this->encoded_matrix.push_back(firstValue);
            
            for(int col = 1; col < width; col++)
            {
                int value = (int)(this->original_matrix.at(col) - predictor->predict(this->original_matrix.at(col-1), 0, 0));
                this->encoded_matrix.push_back(value);
            }
            
            for(int line = 1; line < height; line++)
            {
                for (int col = 0; col < width; col++)
                {   
                    int value;
                    if(col == 0)
                    {
                        value = (int)(this->original_matrix.at(line*width + 0) - predictor->predict(0, this->original_matrix.at((line-1)*width + 0), 0));
                    }
                    else 
                    {   
                        int a = this->original_matrix.at(line*width + (col - 1));
                        int b = this->original_matrix.at((line-1)*width + col);
                        int c = this->original_matrix.at((line-1)*width+ (col - 1));
                        value = (int)(this->original_matrix.at(line*width + col) - predictor->predict(a,b,c));
                    }
                    this->encoded_matrix.push_back(value);
                }
            }

            for(int line = 0; line < height; line++)
            {
                for(int col = 0; col < width; col++)
                {
                    vector<int> code = golomb->encode(this->encoded_matrix.at(line*width + col));
                }
            }
        } 
};

int main() {
    IntraFrameEncoder ife;
    Frame444* frame = new Frame444(720, 1280, "../../../python/media/park_joy_444_720p50.y4m");
    JPEG1* predictor = new JPEG1();
    BitStream* bitstream = new BitStream("../../out/park_joy_444_720p50.bin", 'w');
    Golomb* golomb = new Golomb(4);

    int total_time = 0;
    int total_frames = 0;
    bool firstFrame = true;

    while(true)
    {
        std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();
        bool playing = frame->advance();

        // movie end
        if(!playing)
            break;
        
        vector<int> Y = frame->getY();
        ife.setMatrix(Y);
        if(firstFrame)
        {
            bitstream->writeString("500\t720\t1280");
            firstFrame = false;
        }
            
        ife.encode(frame, predictor, bitstream, golomb);

        vector<int> U = frame->getU();
        ife.setMatrix(U);
        ife.encode(frame, predictor, bitstream, golomb);

        vector<int> V = frame->getV();
        ife.setMatrix(V);
        ife.encode(frame, predictor, bitstream, golomb);
        
        total_frames++;
        std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
        std::cout << "Frame compressed in " << std::chrono::duration_cast<chrono::seconds>(end - begin).count() << " s;\n";
        break;
    }

    return 0;
}